Sources/_CryptoExtras/AES/Nonces.swift

```diff
@@ -0,0 +1,304 @@
+        /// Creates a nonce from the given collection.
+        ///
+        /// Unless your use case calls for a nonce with a specific value, use the
+        /// ``init()`` method to instead create a random nonce.
+        ///
+        /// - Parameters:
+        ///   - nonceBytes: A collection of bytes representation of the nonce. 
+        /// The initializer throws an error if the data has the incorrect length.
+        public init<NonceBytes: Collection>(nonceBytes: NonceBytes) throws where NonceBytes.Element == UInt8 {
+            guard [12, 16].contains(nonceBytes.count) else {
+                throw CryptoKitError.incorrectKeySize
+            }
+
+            var bytes = Self.emptyBytes
+            Swift.withUnsafeMutableBytes(of: &bytes) { bytesPtr in
+                bytesPtr.copyBytes(from: nonceBytes)
+            }
+            self.bytes = bytes
+        }
+        
+        /// Creates a nonce from the given data.
+        ///
+        /// Unless your use case calls for a nonce with a specific value, use the
+        /// ``init()`` method to instead create a random nonce.
+        ///
+        /// - Parameters:
+        ///   - data: A data representation of the nonce. The initializer throws an
+        /// error if the data has the incorrect length.
+        public init<D: DataProtocol>(data: D) throws {
+            guard [12, 16].contains(data.count) else {
+                throw CryptoKitError.incorrectKeySize
+            }
+
+            var bytes = Self.emptyBytes
+            Swift.withUnsafeMutableBytes(of: &bytes) { bytesPtr in
+                data.copyBytes(to: bytesPtr)
+            }
+            self.bytes = bytes
+        }
+        
+        /// Calls the given closure with a pointer to the underlying bytes of the array’s
+        /// contiguous storage.
+        ///
+        /// - Parameters:
+        ///   - body: A closure with an `UnsafeRawBufferPointer` parameter that points to the
+        /// contiguous storage for the array. The system creates the storage if it doesn’t
+        /// exist. If body has a return value, that value is also used as the return value
+        /// for the ``withUnsafeBytes(_:)`` method. The argument is valid only for
+        /// the duration of the closure’s execution.
+        ///
+        /// - Returns: The return value, if any, of the body closure parameter.
+        public func withUnsafeBytes<R>(_ body: (UnsafeRawBufferPointer) throws -> R) rethrows -> R {
+            var bytes = self.bytes
+            return try Swift.withUnsafeBytes(of: &bytes, body)
+        }
+        
+        mutating func withUnsafeMutableBytes<ReturnType>(_ body: (UnsafeMutableRawBufferPointer) throws -> ReturnType) rethrows -> ReturnType {
+            var bytes = self.bytes
+            return try Swift.withUnsafeMutableBytes(of: &bytes, body)
+        }
+        
+        /// Returns an iterator over the elements of the nonce.
+        public func makeIterator() -> Array<UInt8>.Iterator {
```

Comment:
Let's use our own iterator type to hide the face that we're using Array. We _can_ make this cheaper in future, and it would probably be nice to do so.

Reply:
I gave it a shot, and made a generic Iterator that could be used by all three.

It might make sense to bring it into the GBC and ChaChaPoly nonces too